lib/src/cancelable_operation.dart

```diff
@@ -150,6 +150,65 @@ class CancelableOperation<T> {
       {FutureOr<R> Function(Object, StackTrace)? onError,
       FutureOr<R> Function()? onCancel,
       bool propagateCancel = false}) {
+    if (onError != null && onCancel != null) {
+      return thenOperation<R>((value, completer) {
+        completer.complete(onValue(value));
+      }, onError: (error, stack, completer) {
+        completer.complete(onError(error, stack));
+      }, onCancel: (completer) {
+        completer.complete(onCancel());
+      }, propagateCancel: propagateCancel);
+    } else if (onError != null) {
+      return thenOperation<R>((value, completer) {
+        completer.complete(onValue(value));
+      }, onError: (error, stack, completer) {
+        completer.complete(onError(error, stack));
+      }, propagateCancel: propagateCancel);
+    } else if (onCancel != null) {
+      return thenOperation<R>((value, completer) {
+        completer.complete(onValue(value));
+      }, onCancel: (completer) {
+        completer.complete(onCancel());
+      }, propagateCancel: propagateCancel);
+    } else {
+      return thenOperation<R>((value, completer) {
+        completer.complete(onValue(value));
+      }, propagateCancel: propagateCancel);
+    }
+  }
+
+  /// Creates a new cancelable operation to be completed
+  /// when this operation completes or is cancelled.
+  ///
+  /// The [onValue] and [onError] callbacks behave in the same way as
+  /// for [Future.then], and the result of those callbacks is used to complete
```

Comment:
same way ... well, almost. `Future.then` is a little clever. If the `onError` handler throws the *same* error object that it got in, it considers it a rethrow and uses the old stack instead of the new.
I don't actually think that's valuable here, but maybe change "in the same way as" to "similarly to".